File: shipley_checks/src/Stage_4/metrics/m2_growth_compatibility.R
#!/usr/bin/env Rscript
#
# METRIC 2: GROWTH COMPATIBILITY (CSR CONFLICTS)
#
# ============================================================================
# ECOLOGICAL RATIONALE
# ============================================================================
#
# The CSR framework (Grime 1977) classifies plant strategies along 3 axes:
#
# C (COMPETITIVE): Fast-growing, resource-demanding plants that dominate in
#                  high-resource environments. Tall, broad leaves, rapid
#                  biomass accumulation. Examples: trees, large shrubs.
#
# S (STRESS-TOLERANT): Slow-growing, efficient plants adapted to low-resource
#                      environments (drought, poor soil, shade). Conservative
#                      growth, long-lived leaves. Examples: succulents, alpine
#                      plants, shade-tolerant understory.
#
# R (RUDERAL): Short-lived, opportunistic plants that exploit temporary
#              resource pulses. Rapid maturation, high seed production.
#              Examples: annuals, weedy herbs.
#
# CONFLICTS arise when plants with incompatible strategies compete for space
# and resources:
#
# 1. C-C (High-Competitive vs High-Competitive):
#    - Both plants aggressively compete for light, water, nutrients
#    - SEVERITY: 1.0 (base) - highest conflict
#    - MODULATION:
#      * Growth form: Vines on trees = 0.2× (complementary)
#      * Growth form: Trees + herbs = 0.4× (different niches)
#      * Height difference: >5m = 0.3×, 2-5m = 0.6×, <2m = 1.0×
#
# 2. C-S (High-Competitive vs High-Stress-Tolerant):
#    - C plants create shade/resource depletion that S plants cannot tolerate
#    - SEVERITY: 0.6 (base) - moderate conflict
#    - CRITICAL MODULATION: Light preference of S plant
#      * S is shade-adapted (EIVE-L 1-3) → 0.0× (S WANTS shade from C!)
#      * S is sun-loving (EIVE-L 8-9) → 0.9× (C will shade out S)
#      * S is flexible (EIVE-L 4-7) → 0.6× with height modulation
#
# 3. C-R (High-Competitive vs High-Ruderal):
#    - C plants shade and outcompete short-lived R plants
#    - SEVERITY: 0.8 (base) - high conflict
#    - MODULATION: Height difference >5m = 0.3× (R can exploit gaps)
#
# 4. R-R (High-Ruderal vs High-Ruderal):
#    - Both are short-lived, rapid-turnover species
#    - SEVERITY: 0.3 - low conflict (ephemeral, different timing)
#
# ============================================================================
# CALCULATION STEPS
# ============================================================================
#
# STEP 1: Classify plants using PERCENTILE-BASED thresholds
#   - Convert raw CSR scores (0-100) to percentiles using GLOBAL calibration
#   - High-C: C_percentile > 75 (top quartile)
#   - High-S: S_percentile > 75
#   - High-R: R_percentile > 75
#   - WHY GLOBAL? Conflicts are within-guild comparisons, not cross-guild
#
# STEP 2: Detect pairwise conflicts
#   - For each conflict type (C-C, C-S, C-R, R-R), iterate through all pairs
#   - Apply base conflict severity
#   - Apply modulation factors (growth form, height, light preference)
#   - Sum all conflict scores
#
# STEP 3: Normalize by guild size (conflict density)
#   - conflict_density = total_conflicts / max_possible_pairs
#   - max_possible_pairs = n_plants × (n_plants - 1)
#   - This makes scores comparable across guild sizes (2-7 plants)
#
# STEP 4: Percentile normalization
#   - Uses Köppen tier-stratified calibration for metric 'n4'
#   - HIGH conflict_density → HIGH percentile (bad compatibility)
#   - Note: invert = FALSE (no inversion during normalization)
#
# STEP 5: Convert to display score
#   - Final score: 100 - percentile
#   - This inverts the scale so HIGH = GOOD for display
#   - Matches Python guild_scorer_v3.py line 407:
#     'growth_compatibility': 100 - percentiles['n4']
#
# ============================================================================
# PARITY REQUIREMENTS
# ============================================================================
#
# To maintain 100% parity with Python scorer:
# 1. Must use same CSR percentile calibration:
#    shipley_checks/stage4/csr_percentile_calibration_global.json
# 2. Must use same percentile threshold: 75 (top quartile)
# 3. Must apply same modulation factors and weights
# 4. Must use same conflict density normalization
# 5. Must use same Köppen tier calibration file for 'n4' metric
# 6. Must apply same final transformation (100 - percentile)
#
# Python implementation: guild_scorer_v3.py lines 1047-1209
#
# ============================================================================


#' Calculate M2: Growth Compatibility (CSR Conflicts)
#'
#' Scores ecological compatibility based on Grime's CSR strategy conflicts.
#' Detects 4 types of conflicts (C-C, C-S, C-R, R-R) with context-specific
#' modulation based on growth form, height, and light preference.
#'
#' @param guild_plants Data frame with columns:
#'   - CSR_C, CSR_S, CSR_R: Raw CSR scores (0-100)
#'   - height_m: Plant height in meters
#'   - try_growth_form: Growth form (tree, shrub, herb, vine, etc.)
#'   - light_pref: EIVE Light preference (1-9 scale)
#'   - wfo_scientific_name: Plant name (for conflict details)
#' @param csr_to_percentile_fn Function to convert raw CSR to percentile
#' @param percentile_normalize_fn Function to convert raw score to percentile
#'
#' @return List with three elements:
#'   - raw: Conflict density (conflicts per possible pair)
#'   - norm: Normalized percentile (0-100, matches Python before inversion)
#'   - details: List with raw_conflicts, conflict_density, conflict list,
#'              and counts of high-C/S/R plants
#'
#' @details
#' The PERCENTILE_THRESHOLD (75) ensures consistent classification across
#' the full plant database. Plants with C/S/R scores in the top quartile
#' are considered "high" in that strategy.
#'
#' @references
#' Python implementation: src/Stage_4/guild_scorer_v3.py lines 1047-1209
#' Grime, J. P. (1977). Evidence for the existence of three primary strategies
#' in plants and its relevance to ecological and evolutionary theory.
#'
calculate_m2_growth_compatibility <- function(guild_plants,
                                              csr_to_percentile_fn,
                                              percentile_normalize_fn) {

  n_plants <- nrow(guild_plants)
  conflicts <- 0
  conflict_details <- list()

  # -------------------------------------------------------------------------
  # STEP 1: Convert raw CSR scores to GLOBAL percentiles
  # -------------------------------------------------------------------------
  # Why global? CSR conflicts are relative within the guild, not across guilds.
  # A plant with C=60 might be "high-C" in a guild of stress-tolerant plants,
  # but "low-C" in a guild of competitive trees.
  #
  # The percentile transformation provides consistent classification across
  # the entire plant database.

  PERCENTILE_THRESHOLD <- 75  # Top quartile

  guild_plants_copy <- guild_plants
  guild_plants_copy$C_percentile <- sapply(guild_plants_copy$CSR_C, function(x) {
    csr_to_percentile_fn(x, 'c')
  })
  guild_plants_copy$S_percentile <- sapply(guild_plants_copy$CSR_S, function(x) {
    csr_to_percentile_fn(x, 's')
  })
  guild_plants_copy$R_percentile <- sapply(guild_plants_copy$CSR_R, function(x) {
    csr_to_percentile_fn(x, 'r')
  })

  # -------------------------------------------------------------------------
  # CONFLICT TYPE 1: High-C + High-C (Competitive vs Competitive)
  # -------------------------------------------------------------------------
  # ECOLOGICAL CONTEXT:
  # Two fast-growing, resource-demanding plants in the same guild will
  # aggressively compete for light, water, and nutrients. This is the most
  # severe conflict type.
  #
  # MODULATION FACTORS:
  # 1. Vine/Liana + Tree: Vines use trees as structural support without
  #    competing for light (complementary relationship) → 0.2× severity
  #
  # 2. Tree + Herb: Trees occupy canopy, herbs occupy ground layer.
  #    Different niches reduce competition → 0.4× severity
  #
  # 3. Height difference (for plants in same growth form):
  #    - <2m: Same canopy layer, direct competition → 1.0× severity
  #    - 2-5m: Partial niche separation → 0.6× severity
  #    - >5m: Different canopy layers, reduced competition → 0.3× severity
  #
  # Python reference: lines 1104-1141

  high_c_plants <- guild_plants_copy %>% dplyr::filter(C_percentile > PERCENTILE_THRESHOLD)

  if (nrow(high_c_plants) >= 2) {
    for (i in 1:(nrow(high_c_plants) - 1)) {
      for (j in (i + 1):nrow(high_c_plants)) {
        plant_a <- high_c_plants[i, ]
        plant_b <- high_c_plants[j, ]

        conflict <- 1.0  # Base severity

        # Get growth forms (handle NA values)
        form_a <- tolower(as.character(plant_a$try_growth_form))
        form_b <- tolower(as.character(plant_b$try_growth_form))
        form_a <- ifelse(is.na(plant_a$try_growth_form), '', form_a)
        form_b <- ifelse(is.na(plant_b$try_growth_form), '', form_b)

        # MODULATION: Growth form complementarity
        if ((grepl('vine', form_a) || grepl('liana', form_a)) && grepl('tree', form_b)) {
          conflict <- conflict * 0.2  # Vine can climb tree
        } else if ((grepl('vine', form_b) || grepl('liana', form_b)) && grepl('tree', form_a)) {
          conflict <- conflict * 0.2  # Vine can climb tree
        } else if ((grepl('tree', form_a) && grepl('herb', form_b)) ||
                   (grepl('tree', form_b) && grepl('herb', form_a))) {
          conflict <- conflict * 0.4  # Different vertical niches
        } else {
          # MODULATION: Height separation
          height_diff <- abs(plant_a$height_m - plant_b$height_m)
          if (height_diff < 2.0) {
            conflict <- conflict * 1.0  # Same canopy layer
          } else if (height_diff < 5.0) {
            conflict <- conflict * 0.6  # Partial separation
          } else {
            conflict <- conflict * 0.3  # Different canopy layers
          }
        }

        conflicts <- conflicts + conflict

        # Record significant conflicts (severity > 0.2) for diagnostics
        if (conflict > 0.2) {
          conflict_details[[length(conflict_details) + 1]] <- list(
            type = 'C-C',
            severity = conflict,
            plants = c(plant_a$wfo_scientific_name, plant_b$wfo_scientific_name)
          )
        }
      }
    }
  }

  # -------------------------------------------------------------------------
  # CONFLICT TYPE 2: High-C + High-S (Competitive vs Stress-Tolerant)
  # -------------------------------------------------------------------------
  # ECOLOGICAL CONTEXT:
  # Competitive plants create environmental stress (shade, resource depletion)
  # that stress-tolerant plants may or may not tolerate depending on their
  # adaptations.
  #
  # CRITICAL MODULATION: Light preference of S plant (EIVE-L scale 1-9)
  #
  # EIVE-L INTERPRETATION:
  # 1-3: Deep shade to moderate shade (< 10% full sunlight)
  # 4-7: Semi-shade to half-light (10-50% full sunlight)
  # 8-9: Full light (> 50% full sunlight)
  #
  # LOGIC:
  # - S is shade-adapted (L < 3.2): S plant WANTS to grow under C plant's
  #   canopy. This is a BENEFICIAL relationship, not a conflict → 0.0× severity
  #
  # - S is sun-loving (L > 7.47): S plant requires full sun. C plant will
  #   shade it out, causing severe stress → 0.9× severity
  #
  # - S is flexible (L 3.2-7.47): S plant can tolerate partial shade.
  #   Conflict depends on height separation:
  #   * >8m height difference: C provides overhead canopy, S thrives below → 0.3×
  #   * Otherwise: base conflict of 0.6×
  #
  # Python reference: lines 1143-1177

  high_s_plants <- guild_plants_copy %>% dplyr::filter(S_percentile > PERCENTILE_THRESHOLD)

  for (idx_c in which(guild_plants_copy$C_percentile > PERCENTILE_THRESHOLD)) {
    plant_c <- guild_plants_copy[idx_c, ]
    for (idx_s in which(guild_plants_copy$S_percentile > PERCENTILE_THRESHOLD)) {
      if (idx_c != idx_s) {
        plant_s <- guild_plants_copy[idx_s, ]

        conflict <- 0.6  # Base severity

        # Get S plant's light preference
        s_light <- plant_s$light_pref

        if (is.na(s_light)) {
          s_light <- 5.0  # Default: flexible (mid-range)
        }

        # CRITICAL: Light-based modulation
        if (s_light < 3.2) {
          # S is SHADE-ADAPTED (L=1-3: deep shade to moderate shade)
          # The S plant WANTS to be under the C plant's canopy!
          # This is a BENEFICIAL relationship, not a conflict
          conflict <- 0.0
        } else if (s_light > 7.47) {
          # S is SUN-LOVING (L=8-9: full-light plant)
          # The C plant will shade it out, causing severe stress
          conflict <- 0.9
        } else {
          # S is FLEXIBLE (L=4-7: semi-shade to half-light)
          # MODULATION: Height difference
          # If C plant is much taller (>8m), S plant can thrive below
          height_diff <- abs(plant_c$height_m - plant_s$height_m)
          if (height_diff > 8.0) {
            conflict <- conflict * 0.3  # Beneficial vertical niche separation
          }
          # Otherwise: use base conflict of 0.6
        }

        conflicts <- conflicts + conflict

        if (conflict > 0.2) {
          conflict_details[[length(conflict_details) + 1]] <- list(
            type = 'C-S',
            severity = conflict,
            plants = c(plant_c$wfo_scientific_name, plant_s$wfo_scientific_name)
          )
        }
      }
    }
  }

  # -------------------------------------------------------------------------
  # CONFLICT TYPE 3: High-C + High-R (Competitive vs Ruderal)
  # -------------------------------------------------------------------------
  # ECOLOGICAL CONTEXT:
  # Competitive plants create long-term resource monopolies that short-lived
  # ruderal plants cannot tolerate. Ruderals need open, disturbed habitats
  # with temporary resource pulses.
  #
  # MODULATION:
  # - Height difference >5m: R plants can exploit gaps and edges around
  #   C plants before canopy closure → 0.3× severity
  # - Otherwise: High conflict → 0.8× severity (base)
  #
  # Python reference: lines 1179-1199

  high_r_plants <- guild_plants_copy %>% dplyr::filter(R_percentile > PERCENTILE_THRESHOLD)

  for (idx_c in which(guild_plants_copy$C_percentile > PERCENTILE_THRESHOLD)) {
    plant_c <- guild_plants_copy[idx_c, ]
    for (idx_r in which(guild_plants_copy$R_percentile > PERCENTILE_THRESHOLD)) {
      if (idx_c != idx_r) {
        plant_r <- guild_plants_copy[idx_r, ]

        conflict <- 0.8  # Base severity

        # MODULATION: Height difference
        # Tall C plants create gaps where R plants can complete their
        # rapid life cycle before canopy closure
        height_diff <- abs(plant_c$height_m - plant_r$height_m)
        if (height_diff > 5.0) {
          conflict <- conflict * 0.3  # Temporal niche separation
        }

        conflicts <- conflicts + conflict

        if (conflict > 0.2) {
          conflict_details[[length(conflict_details) + 1]] <- list(
            type = 'C-R',
            severity = conflict,
            plants = c(plant_c$wfo_scientific_name, plant_r$wfo_scientific_name)
          )
        }
      }
    }
  }

  # -------------------------------------------------------------------------
  # CONFLICT TYPE 4: High-R + High-R (Ruderal vs Ruderal)
  # -------------------------------------------------------------------------
  # ECOLOGICAL CONTEXT:
  # Both plants are short-lived, opportunistic species with rapid turnover.
  # They typically have different timing (spring ephemerals vs summer annuals)
  # or occupy slightly different microsites within disturbed patches.
  #
  # SEVERITY: 0.3 - Low conflict (temporal/spatial niche separation)
  #
  # Python reference: lines 1201-1206

  if (nrow(high_r_plants) >= 2) {
    for (i in 1:(nrow(high_r_plants) - 1)) {
      for (j in (i + 1):nrow(high_r_plants)) {
        conflict <- 0.3  # Low severity - ephemeral species
        conflicts <- conflicts + conflict

        # Note: We don't record R-R conflicts in details (always low severity)
      }
    }
  }

  # -------------------------------------------------------------------------
  # STEP 3: Normalize by guild size (conflict density)
  # -------------------------------------------------------------------------
  # Conflict density = total conflicts / maximum possible pairs
  #
  # This normalization ensures scores are comparable across guild sizes:
  # - 2 plants: max 2 pairs (A→B, B→A)
  # - 3 plants: max 6 pairs (A→B, A→C, B→A, B→C, C→A, C→B)
  # - 7 plants: max 42 pairs
  #
  # Without this normalization, larger guilds would always have higher
  # conflict scores simply due to more pairwise comparisons.

  max_pairs <- if (n_plants > 1) n_plants * (n_plants - 1) else 1
  conflict_density <- conflicts / max_pairs

  # -------------------------------------------------------------------------
  # STEP 4: Percentile normalization using Köppen tier calibration
  # -------------------------------------------------------------------------
  # Converts absolute conflict density to relative percentile within the
  # calibration dataset for the specified Köppen climate tier.
  #
  # Metric name: 'n4' (legacy naming from earlier versions)
  # invert = FALSE: HIGH conflict_density → HIGH percentile (bad compatibility)

  m2_norm <- percentile_normalize_fn(conflict_density, 'n4', invert = FALSE)

  # -------------------------------------------------------------------------
  # RETURN: Raw conflict density, normalized percentile, and diagnostics
  # -------------------------------------------------------------------------

  list(
    raw = conflict_density,
    norm = m2_norm,
    details = list(
      raw_conflicts = conflicts,
      conflict_density = conflict_density,
      conflicts = conflict_details,  # List of significant conflicts for debugging
      high_c = nrow(high_c_plants),
      high_s = nrow(high_s_plants),
      high_r = nrow(high_r_plants)
    )
  )
}

File: shipley_checks/src/Stage_4/guild_scorer_rust/src/metrics/m2_growth_compatibility.rs
//! METRIC 2: GROWTH COMPATIBILITY (CSR CONFLICTS)
//!
//! **PHASE 2 OPTIMIZATION**: LazyFrame with column projection
//!
//! Scores ecological compatibility based on Grime's CSR strategy conflicts.
//! Detects 4 types of conflicts (C-C, C-S, C-R, R-R) with context-specific
//! modulation based on growth form, height, and light preference.
//!
//! **Memory optimization**:
//!   - Old: Takes &DataFrame with ALL 782 columns materialized
//!   - New: Takes &LazyFrame and materializes ONLY 7 needed columns
//!   - Savings: 7 plants × 782 cols = 5,474 cells → 7 plants × 7 cols = 49 cells
//!   - **111× less data loaded into memory**
//!
//! **Columns needed** (only these 7):
//!   1. wfo_scientific_name  - Plant identification
//!   2. CSR_C                - Competitor score (raw)
//!   3. CSR_S                - Stress-tolerator score (raw)
//!   4. CSR_R                - Ruderal score (raw)
//!   5. height_m             - Plant height for vertical niche analysis
//!   6. try_growth_form      - Growth form (tree, herb, vine) for conflict modulation
//!   7. light_pref           - Light preference (EIVE-L) for C-S conflict assessment
//!
//! R reference: shipley_checks/src/Stage_4/metrics/m2_growth_compatibility.R

use polars::prelude::*;
use anyhow::{Result, Context};
use crate::utils::{Calibration, CsrCalibration, percentile_normalize, csr_to_percentile, filter_to_guild};
use std::collections::HashSet;

const PERCENTILE_THRESHOLD: f64 = 75.0; // Top quartile

/// Column requirements for M2 calculation
pub const REQUIRED_PLANT_COLS: &[&str] = &[
    "wfo_taxon_id",         // WFO ID for filtering
    "wfo_scientific_name",  // Plant name for diagnostics
    "CSR_C",                // Competitor raw score (aliased from "C")
    "CSR_S",                // Stress-tolerator raw score (aliased from "S")
    "CSR_R",                // Ruderal raw score (aliased from "R")
    "height_m",             // Height for vertical niche analysis
    "try_growth_form",      // Growth form (tree/herb/vine)
    "light_pref",           // Light preference (aliased from "EIVEres-L_complete")
];

/// Per-plant CSR data for detailed breakdown
#[derive(Debug, Clone)]
pub struct PlantCsrData {
    pub plant_name: String,
    pub display_name: String,
    pub c_raw: f64,
    pub s_raw: f64,
    pub r_raw: f64,
    pub c_percentile: f64,
    pub s_percentile: f64,
    pub r_percentile: f64,
    pub dominant_strategy: String,
}

/// Result of M2 calculation
#[derive(Debug, Clone)]
pub struct M2Result {
    /// Conflict density (conflicts per possible pair)
    pub raw: f64,
    /// Normalized percentile (0-100)
    pub norm: f64,
    /// Number of high-C plants
    pub high_c_count: usize,
    /// Number of high-S plants
    pub high_s_count: usize,
    /// Number of high-R plants
    pub high_r_count: usize,
    /// Total raw conflicts before density normalization
    pub total_conflicts: f64,
    /// Per-plant CSR data for detailed breakdown
    pub plant_csr_data: Vec<PlantCsrData>,
}

/// Plant data for conflict detection
#[derive(Debug, Clone)]
#[allow(dead_code)] // Fields used for debugging/diagnostics
struct PlantRow {
    index: usize,
    name: String,
    display_name: String,  // With vernacular (e.g., "Vitis vinifera (Common Grape)")
    csr_c: f64,
    csr_s: f64,
    csr_r: f64,
    c_percentile: f64,
    s_percentile: f64,
    r_percentile: f64,
    height_m: f64,
    growth_form: String,
    light_pref: f64,
}

/// Calculate M2: Growth Compatibility (CSR Conflicts)
///
/// **OPTIMIZATION**: Uses LazyFrame with projection pruning
///
/// **Old approach** (pre-Phase 2):
///   ```
///   let guild_plants = plants_df.filter(&mask)?;  // Loads ALL 782 columns
///   let m2 = calculate_m2(&guild_plants, ...)?;
///   ```
///   - Memory: 7 plants × 782 cols = 5,474 cells in RAM
///
/// **New approach** (Phase 2):
///   ```
///   let m2 = calculate_m2(&plants_lazy, plant_ids, ...)?;
///   ```
///   - Inside this function, Polars:
///     1. Filters during Parquet scan (predicate pushdown)
///     2. Loads ONLY 7 columns (projection pruning)
///   - Memory: 7 plants × 7 cols = 49 cells in RAM
///   - **111× less data movement**
///
/// R reference: m2_growth_compatibility.R::calculate_m2_growth_compatibility
pub fn calculate_m2(
    plants_lazy: &LazyFrame,         // Schema-only LazyFrame (not materialized)
    plant_ids: &[String],            // WFO IDs to filter for this guild
    csr_calibration: Option<&CsrCalibration>,
    calibration: &Calibration,
) -> Result<M2Result> {
    // ========================================================================
    // STEP 1: Materialize columns with aliasing (M2-specific)
    // ========================================================================
    //
    // M2 requires aliasing of parquet columns (C->CSR_C, S->CSR_S, etc.)
    // so we handle materialization inline but use helper for filtering

    let plants_filtered = plants_lazy
        .clone()
        .select(&[
            col("wfo_taxon_id"),
            col("wfo_scientific_name"),
            col("vernacular_name_en"),  // For display names
            col("C").alias("CSR_C"),
            col("S").alias("CSR_S"),
            col("R").alias("CSR_R"),
            col("height_m"),
            col("try_growth_form"),
            col("EIVEres-L_complete").alias("light_pref"),
        ])
        .collect()
        .with_context(|| "M2: Failed to materialize plant columns")?;

    // VALIDATE: Check all expected columns present
    let actual_cols: HashSet<String> = plants_filtered.get_column_names()
        .into_iter()
        .map(|s| s.to_string())
        .collect();

    for &expected in REQUIRED_PLANT_COLS {
        if !actual_cols.contains(expected) {
            anyhow::bail!(
                "M2: Missing expected column '{}'. Available columns: {:?}",
                expected, actual_cols
            );
        }
    }

    // ========================================================================
    // STEP 2: Filter to guild plants using helper
    // ========================================================================

    let guild_plants = filter_to_guild(&plants_filtered, plant_ids, "wfo_taxon_id", "M2")?;

    let n_plants = guild_plants.height();

    // ========================================================================
    // STEP 3: Extract data and calculate conflicts
    // ========================================================================
    //
    // This part is unchanged - same conflict detection logic as before
    // But now operating on a much smaller DataFrame (49 cells vs 5,474 cells)

    let plants = extract_plant_data(&guild_plants, csr_calibration)?;

    // Classify plants
    let high_c: Vec<&PlantRow> = plants.iter()
        .filter(|p| p.c_percentile > PERCENTILE_THRESHOLD)
        .collect();
    let high_s: Vec<&PlantRow> = plants.iter()
        .filter(|p| p.s_percentile > PERCENTILE_THRESHOLD)
        .collect();
    let high_r: Vec<&PlantRow> = plants.iter()
        .filter(|p| p.r_percentile > PERCENTILE_THRESHOLD)
        .collect();

    let mut total_conflicts = 0.0;

    // CONFLICT TYPE 1: C-C (Competitive vs Competitive)
    if high_c.len() >= 2 {
        for i in 0..high_c.len() - 1 {
            for j in i + 1..high_c.len() {
                let conflict = calculate_c_c_conflict(high_c[i], high_c[j]);
                total_conflicts += conflict;
            }
        }
    }

    // CONFLICT TYPE 2: C-S (Competitive vs Stress-Tolerant)
    for plant_c in &high_c {
        for plant_s in &high_s {
            if plant_c.index != plant_s.index {
                let conflict = calculate_c_s_conflict(plant_c, plant_s);
                total_conflicts += conflict;
            }
        }
    }

    // CONFLICT TYPE 3: C-R (Competitive vs Ruderal)
    for plant_c in &high_c {
        for plant_r in &high_r {
            if plant_c.index != plant_r.index {
                let conflict = calculate_c_r_conflict(plant_c, plant_r);
                total_conflicts += conflict;
            }
        }
    }

    // CONFLICT TYPE 4: R-R (Ruderal vs Ruderal)
    if high_r.len() >= 2 {
        for i in 0..high_r.len() - 1 {
            for _j in i + 1..high_r.len() {
                total_conflicts += 0.3; // Fixed low severity
            }
        }
    }

    // Normalize by guild size (conflict density)
    let max_pairs = if n_plants > 1 {
        n_plants * (n_plants - 1)
    } else {
        1
    };
    let conflict_density = total_conflicts / max_pairs as f64;

    // Percentile normalization (Köppen tier-stratified)
    let m2_norm = percentile_normalize(conflict_density, "n4", calibration, false)?;

    // Build per-plant CSR data for detailed breakdown
    let plant_csr_data: Vec<PlantCsrData> = plants.iter().map(|p| {
        PlantCsrData {
            plant_name: p.name.clone(),
            display_name: p.display_name.clone(),
            c_raw: p.csr_c,
            s_raw: p.csr_s,
            r_raw: p.csr_r,
            c_percentile: p.c_percentile,
            s_percentile: p.s_percentile,
            r_percentile: p.r_percentile,
            dominant_strategy: determine_dominant_strategy(
                p.c_percentile,
                p.s_percentile,
                p.r_percentile
            ),
        }
    }).collect();

    Ok(M2Result {
        raw: conflict_density,
        norm: m2_norm,
        high_c_count: high_c.len(),
        high_s_count: high_s.len(),
        high_r_count: high_r.len(),
        total_conflicts,
        plant_csr_data,
    })
}

/// Determine dominant CSR strategy based on percentiles
///
/// Returns the strategy with the highest percentile, or "Mixed" if balanced
fn determine_dominant_strategy(c_pct: f64, s_pct: f64, r_pct: f64) -> String {
    // Check if strategies are relatively balanced (within 20 percentile points)
    let max_pct = c_pct.max(s_pct).max(r_pct);
    let min_pct = c_pct.min(s_pct).min(r_pct);

    if max_pct - min_pct < 20.0 {
        return "Mixed".to_string();
    }

    // Otherwise return the dominant strategy
    if c_pct >= s_pct && c_pct >= r_pct {
        if c_pct > PERCENTILE_THRESHOLD {
            "Competitive".to_string()
        } else {
            "C-leaning".to_string()
        }
    } else if s_pct >= c_pct && s_pct >= r_pct {
        if s_pct > PERCENTILE_THRESHOLD {
            "Stress-tolerant".to_string()
        } else {
            "S-leaning".to_string()
        }
    } else {
        if r_pct > PERCENTILE_THRESHOLD {
            "Ruderal".to_string()
        } else {
            "R-leaning".to_string()
        }
    }
}

/// Extract plant data from DataFrame and convert CSR to percentiles
fn extract_plant_data(
    df: &DataFrame,
    csr_calibration: Option<&CsrCalibration>,
) -> Result<Vec<PlantRow>> {
    use crate::utils::get_display_name;

    let n = df.height();
    let mut plants = Vec::with_capacity(n);

    let names = df.column("wfo_scientific_name")?.str()?;
    let vernacular_en = df.column("vernacular_name_en").ok().and_then(|c| c.str().ok());
    let csr_c = df.column("CSR_C")?.f64()?;
    let csr_s = df.column("CSR_S")?.f64()?;
    let csr_r = df.column("CSR_R")?.f64()?;
    let heights = df.column("height_m")?.f64()?;
    let growth_forms = df.column("try_growth_form")?.str()?;
    let light_prefs = df.column("light_pref")?.f64()?;

    for i in 0..n {
        // Check for missing CSR values - if any are None, skip this guild
        // Defaulting to 50.0 would distort conflict detection (Issue #NA-handling)
        let name = names.get(i).unwrap_or("Unknown").to_string();

        let c_val = csr_c.get(i).ok_or_else(|| {
            anyhow::anyhow!("Plant {} has missing CSR_C data - cannot calculate M2", name)
        })?;
        let s_val = csr_s.get(i).ok_or_else(|| {
            anyhow::anyhow!("Plant {} has missing CSR_S data - cannot calculate M2", name)
        })?;
        let r_val = csr_r.get(i).ok_or_else(|| {
            anyhow::anyhow!("Plant {} has missing CSR_R data - cannot calculate M2", name)
        })?;

        // Build display name with vernacular
        let en = vernacular_en.and_then(|col| col.get(i));
        let display_name = get_display_name(&name, en, None);

        plants.push(PlantRow {
            index: i,
            name,
            display_name,
            csr_c: c_val,
            csr_s: s_val,
            csr_r: r_val,
            c_percentile: csr_to_percentile(c_val, 'c', csr_calibration),
            s_percentile: csr_to_percentile(s_val, 's', csr_calibration),
            r_percentile: csr_to_percentile(r_val, 'r', csr_calibration),
            height_m: heights.get(i).unwrap_or(1.0),
            growth_form: growth_forms.get(i).unwrap_or("").to_lowercase(),
            light_pref: light_prefs.get(i).unwrap_or(5.0),
        });
    }

    Ok(plants)
}

/// Calculate C-C conflict with growth form and height modulation
///
/// R reference: m2_growth_compatibility.R lines 187-235
fn calculate_c_c_conflict(plant_a: &PlantRow, plant_b: &PlantRow) -> f64 {
    let mut conflict = 1.0; // Base severity

    // Growth form complementarity
    let form_a = &plant_a.growth_form;
    let form_b = &plant_b.growth_form;

    if (form_a.contains("vine") || form_a.contains("liana")) && form_b.contains("tree") {
        conflict *= 0.2; // Vine can climb tree
    } else if (form_b.contains("vine") || form_b.contains("liana")) && form_a.contains("tree") {
        conflict *= 0.2; // Vine can climb tree
    } else if (form_a.contains("tree") && form_b.contains("herb"))
        || (form_b.contains("tree") && form_a.contains("herb"))
    {
        conflict *= 0.4; // Different vertical niches
    } else {
        // Height separation
        let height_diff = (plant_a.height_m - plant_b.height_m).abs();
        if height_diff < 2.0 {
            conflict *= 1.0; // Same canopy layer
        } else if height_diff < 5.0 {
            conflict *= 0.6; // Partial separation
        } else {
            conflict *= 0.3; // Different canopy layers
        }
    }

    conflict
}

/// Calculate C-S conflict with critical light preference modulation
///
/// R reference: m2_growth_compatibility.R lines 268-315
fn calculate_c_s_conflict(plant_c: &PlantRow, plant_s: &PlantRow) -> f64 {
    let mut conflict = 0.6; // Base severity

    let s_light = plant_s.light_pref;

    // Critical: Light-based modulation
    if s_light < 3.2 {
        // S is SHADE-ADAPTED - wants to be under C plant's canopy
        conflict = 0.0;
    } else if s_light > 7.47 {
        // S is SUN-LOVING - will be shaded out by C plant
        conflict = 0.9;
    } else {
        // S is FLEXIBLE - depends on height difference
        let height_diff = (plant_c.height_m - plant_s.height_m).abs();
        if height_diff > 8.0 {
            conflict *= 0.3; // Beneficial vertical niche separation
        }
        // Otherwise use base conflict of 0.6
    }

    conflict
}

/// Calculate C-R conflict with height modulation
///
/// R reference: m2_growth_compatibility.R lines 334-361
fn calculate_c_r_conflict(plant_c: &PlantRow, plant_r: &PlantRow) -> f64 {
    let mut conflict = 0.8; // Base severity

    // Height difference allows R plants to exploit gaps
    let height_diff = (plant_c.height_m - plant_r.height_m).abs();
    if height_diff > 5.0 {
        conflict *= 0.3; // Temporal niche separation
    }

    conflict
}
